### ISSUE TITLE
[CFE] Incorrect normalization of Never* to Null in constants in weak mode

### ISSUE THREAD

**Issue by alexmarkov** [dart-lang/sdk#44246]
Consider the following code:

foo.dart
```dart
// @dart=2.9

import 'opted_in.dart' show optedInListOfNever, optedInListOfNullableNever;

const optedOutListOfNever = const <Never>[];

main(List<String> args) {
 print(optedOutListOfNever.runtimeType);
 print(optedInListOfNever.runtimeType);
 print(optedInListOfNullableNever.runtimeType);
}
```

opted_in.dart
```dart
// @dart=2.12

const optedInListOfNever = const <Never>[];
const optedInListOfNullableNever = const <Never?>[];
```

According to https://github.com/dart-lang/language/blob/master/accepted/future-releases/nnbd/feature-specification.md#const-evaluation-and-canonicalization in the weak mode all types in constants should be treated as legacy. This results in `Never*` type argument in `optedOutListOfNever` and `optedInListOfNever` constants.

Then, https://github.com/dart-lang/language/blob/master/resources/type-system/normalization.md does not have a rule to normalize `Never*` to `Null` (only `Never?` is normalized to `Null`).

However, CFE produces the following kernel for this example in the weak mode:
```
library from "foo.dart" as foo {

  import "opted_in.dart" show optedInListOfNever, optedInListOfNullableNever;

  static const field core::List<Null>* optedOutListOfNever = #C1;
  static method main(core::List<core::String*>* args) → dynamic {
    core::print((#C1).{core::Object::runtimeType});
    core::print((#C1).{core::Object::runtimeType});
    core::print((#C1).{core::Object::runtimeType});
  }
}
library from "opted_in.dart" as opt {

  static const field core::List<Never> optedInListOfNever = #C1;
  static const field core::List<Never?> optedInListOfNullableNever = #C1;
}

...

constants  {
  #C1 = <Null>[]

```

All 3 constants have `Null` type argument. If I understand the specifications correctly, `optedOutListOfNever` and `optedInListOfNever` should have `Never*` type argument, and only `optedInListOfNullableNever` should have `Null` type argument.

@leafpetersen @johnniwinther @stefantsov

**Comment by johnniwinther** [comment]
For lack of a definition of "be treated as legacy" the CFE uses `LEGACY_ERASURE`. I believe that @leafpetersen was going to give a precise specification of what to do.

**Comment by alexmarkov** [comment]
Yeah, we definitely need more precise wording in the spec. My reading was that every type `T` is converted to legacy `T*`.

We're experimenting with stricter null safety type checks (under the flag) even in weak mode, in order to be able to catch more bugs without migrating the whole application. Unfortunately, legacy erasure, if applied to types in constants in opted-in libraries breaks subtyping within opted-in code (even without involving opted-out code): `Never` (and `Never*`) are assignable to non-nullable types, but `Null` isn't.

Can we literally make all types legacy in constants (`T` => `T*` recursively for every type T), or make legacy erasure less aggressive so it would convert `Never` to `Never*` and not `Null`?

**Comment by johnniwinther** [comment]
@leafpetersen What is the way forward here?

**Comment by johnniwinther** [comment]
ping @leafpetersen

**Comment by johnniwinther** [comment]
@leafpetersen what should we do here?

**Comment by leafpetersen** [comment]
Sorry, this is one of the areas I've been meaning to come back to during beta.  Agreed that the spec is too vague here, as witnessed by the fact that the analyzer and the CFE take entirely different interpretations here.  In particular:
  1) The CFE interprets "treated as if all type arguments passed to them were legacy types" as meaning that `?` is erased, where the analyzer does not.
      - e.g. the  CFE treats `const <int>[]` and `const<int?>[]` as identical in weak mode, but the analyzer does not
  2) The CFE interprets the same passage as implying that `Never` is treated as `Null`, where the analyzer does not.
      - e.g. the CFE treats `const <Never>[]` and `const <Null>[]` as identical, but the analyzer does not.

My goal here was that constants which are spelled the same should canonicalize as identical, regardless of the library in which they occur.  Hence, the CFE behavior in 1) is not really what I intended here, though it is a reasonable interpretation.

The issue in 2) is more difficult.  While `const <Never>[]` and `const <Null>[]` are spelled differently, `const []` will be interpreted as the former in opted in code, and the latter in opted out code, which is potentially problematic.  @scheglov  raises the specific issue of the following:

```
class A<T> {
  void foo({List<T> a = const []}) {}
}

// Legacy
class B<T> extends A<T> {
  void foo({List<T> a = const []}) {}
}
```

This code is invalid under the current analyzer interpretation, but valid under the CFE interpretation.

I believe that the right interpretation is to take the analyzer's interpretation of 1), but the CFEs interpretation of 2).  That is, we say that all type arguments to const objects are subject to a rewrite (in weak mode) in which a `*` is **added** to every type (recursively) through the structure of types, and every occurrence of `Never` is treated as `Null`.

Does this seem reasonable?   Issues with this?

@scheglov I realize that the lack of a weak/strong mode distinction in the analyzer means that we may have to make some compromises on fidelity here.

cc @eernstg @lrhn

**Comment by johnniwinther** [comment]
I see no problem in implementing this in the CFE.

**Comment by eernstg** [comment]
> a `*` is added to every type (recursively)

I proposed the same thing in https://github.com/dart-lang/language/issues/1346, but:

>  ... and every occurrence of `Never` is treated as `Null`.

... with respect to `Never` I proposed that it should not be an exception (so we add a `*` and get `Never*`, rather than transforming it into `Null`). Any occurrence of `Never` in a legacy library would still be turned into `Null`.

The reason why we might want to do that is that we'd otherwise need to have `--sound-null-safety` respectively `--no-sound-null-safety` modes in the analyzer, and I assume that we don't want that.

Consider the situation we'd have if we do turn `Never` into `Null` when it is passed as or in a type argument in a constant collection literal or constant object expression:

```dart
// This library enables null safety.

const xs = <Never>[];
const ys = <Null>[];
const z = identical(xs, ys)? 1 ~/ 0 : 0;

void main() {}
```

The point is that the initializing expression in the declaration of `z` is a compile-time error if we turn the type argument `Never` into `Null`, that is, with `--no-sound-null-safety`, but it is not a compile-time error with `--sound-null-safety`. So the static analysis cannot be performed unless we know which mode will be used for the execution.

So I'm proposing that we instead turn `Never` into `Never*` when it occurs in an opted-in library and is passed as or in a type argument in a constant collection literal or constant object expression.

This means that, in a mixed version program (where `--no-sound-null-safety` is implied), we can have constant expressions _c1_ (in opted-in code) and _c2_ (in legacy code) that are spelled the same, and still don't get canonicalized to a single object. However, this will _only_ affect expressions where `Never` occurs (written or inferred) in _c1_.
